streams/src/main/java/org/apache/kafka/streams/state/internals/buffer/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -30,6 +31,11 @@
 class InMemoryTimeOrderedKeyValueBuffer implements TimeOrderedKeyValueBuffer {
     private final Map<Bytes, TimeKey> index = new HashMap<>();
     private final TreeMap<TimeKey, ContextualRecord> sortedMap = new TreeMap<>();
+
+    // track keys that need to be flushed.
+    // keys should be removed before update to maintain update order, rather than insertion order.
+    private final LinkedHashSet<Bytes> dirtyKeys = new LinkedHashSet<>();
```

Comment:
Regarding the comment, LinkedHashSet doesn't relocate a key in the linked list when you update its value. It doesn't *strictly* matter for correctness (since there is no ordering between keys). It seemed nice to preserve the relative order of buffer events, so that reprocessing would produce the same sequence of emissions.